docs/rfcs/0001-image-watcher.md

```diff
@@ -0,0 +1,82 @@
+- Start Date: 2020-11-18
+- Target Version: 1.0
+
+# Summary
+This RFC proposes adding a new feature to trigger Deployment when an image tag stored at a container registry is updated.
+Similar to Flux's [Image Scan](https://docs.fluxcd.io/en/1.19.0/faq/#how-often-does-flux-check-for-new-images).
+
+# Motivation
+The canonical deployment flow with PipeCD is:
+
+1. CI server pushes the generated image to the container registry after app-repo updated.
+2. User updates the config-repo manually.
+
+It is the User's responsibility to automate these 1 and 2 step to be done in a series of actions, while it is a bit quite painful.
+Therefore, we would like to solve this problem by letting Piped monitor the container registry.
+
+# Detailed design
+Piped compares the image tags defined in git with the latest tags stored in the container registry, and then pushes them to git if there are any deviations.
+
+### Usage
+Note: Where, we call the client that accesses the container registry to `Image Provider`.
+
+First of all, we need to register the credentials for accessing the image provider with Piped.
+
+```yaml
+apiVersion: pipecd.dev/v1beta1
+kind: Piped
+spec:
+  imageProviders:
+    - name: my-dockerhub
+      type: DOCKERHUB
+      # Indicates how many minutes Piped polls every.
+      interval: 5m
+      config:
+        username: foo
+        passwordFile: /etc/piped-secret/dockerhub-pass
+```
+
+We define the information about image tag underneath `.pipe/` directory:
+```yaml
+apiVersion: pipecd.dev/v1beta1
+kind: ImageWatcher
+spec:
+  targets:
+    - image: gcr.io/pipecd/foo
+      provider: my-gcr
+      filePath: foo/deployment.yaml
+      field: spec.containers[0].image
+    - image: pipecd/bar
+      provider: my-dockerhub
+      filePath: bar/deployment.yaml
+      field: spec.containers[0].image
+```
+
+### Architecture
+`Image Provider` pulls respectively periodically.
+`Image Watcher` pulls from the git repo, compares respectively and pushes the one with differences.
+
+![](assets/image-watcher.jpg)
+
+
+# Alternatives
+Initialy, I was thinking it should be defined in Deployment Configuration, but it's slightly different from configuration about Deployment.
+In addition to that, it is too complicated to inspect the entire git path every time.
+
+```yaml
+apiVersion: pipecd.dev/v1beta1
+kind: KubernetesApp
+spec:
+  imageWatcher:
+    targets:
+    - image: gcr.io/pipecd/helloworld
+      path:
+        filename: deployment.yaml
+        field: spec.containers[0].image
+      provider: my-dockerhub
+```
+
+# Unresolved questions
```

Comment:
Any thoughts about dealing with the rate limit from the container registry?
(caching, grouping, or triggering by CI event...)

Reply:
Yeah, that's what I'm most worried about. The rate limit per patient may be reached when there are a bunch of target images for a single Image Provider.

I know it's a terrible way but I'm going to call `time.Sleep()` at the end of each pull.

https://github.com/pipe-cd/pipe/blob/bff58081a707cf01e326540448912523b2a4f639/pkg/app/piped/imagewatcher/watcher.go#L157-L161

Reply:
The above snippet is in the branch to be deleted after merged, so it will be invisible.

Reply:
Does any provider have a `watch` API or something like that?

Reply:
As far as I investigate, no. Most providers implement the [Docker Registry API](https://docs.docker.com/registry/spec/api/) that has just simple APIs.

It could be a frequent problem, so I'm going to investigate further to see if there's anything else we can do to help.

Reply:
I figured out that no API to watch images, so using the [list tags API](https://docs.docker.com/registry/spec/api/#listing-image-tags) is the only way.

To completely solve the rate limit problem, I re-examined that Piped just provides an API to register in the container registry hooks. But I rejected it because of the heavy burden on users and the need to publish piped.

Therefore, I'm thinking about designing it so that each user takes appropriate action in the case that it may reach the rate limit. For instance, get them to register the premium plan user to Piped as an image provider, or increase the interval between polls).

Reply:
FYI: Let's say there are 100 images to watch in a single Docker hub organization and we want to pull every 5 minutes. With the current design, 100 requests are issued every five minutes.

Reply:
Thanks for your investigation.

Actually, these were what I thought.

- At the initial phase, we do polling as usual.
- In the next phase, we provide a FAKE container registry at our control-plane and add a new kind of IMAGE_PROVIDER for the Piped to use. So for the big projects, they can
     - use our CLI to send metadata about newly created images to our FAKE container registry
     - change their pipeds to our IMAGE_PROVIDER kind

(Out FAKE container registry only contains the metadata of the images)

Reply:
Wow, what a brilliant idea! For the large teams, they use the CLI tool in CI or something like that when their pushing is complete, doesn't it?

Reply:
Okay, I'd be happy to work on the initial phase ;)

Reply:
And make the RFC clear what each step does.

Reply:
Yes, they use our CLI to publish the metadata for their newly created images.
For example, from their CI after pushing the image:

```
pipecd cli push-image --image=abc:v1.0.0 --api-key=key
```